File: java/src/main/java/utils/IoUtils.java
package utils;

import model.PaperSlip;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public final class IoUtils {

    private IoUtils() { }

    public static void outputToCsv(String fileName, List<String> headers, List<List<String>> rows) {

        List<String> lines = Stream.concat(
                Stream.of(String.join(",", headers)),
                rows.stream().map(row -> String.join(",", row))
        ).collect(Collectors.toList());
        try {
            Path outputPath = Path.of("output", fileName + ".csv");
            if (!Files.exists(outputPath.getParent())) {
                Files.createDirectories(outputPath.getParent());
            }
            Files.write(outputPath, lines);
        } catch (IOException e) {
            throw new RuntimeException("Error writing to file", e);
        }
    }

    public static List<String> readExpected(int number) {
        return readFile("answers", String.valueOf(number));
    }

    public static List<PaperSlip> readInput(int number) {
        return readFile("input", String.valueOf(number)).stream().map(PaperSlip::parse).collect(Collectors.toList());
    }

    private static List<String> readFile(String dir, String fileName) {
        try {
            return Files.readAllLines(Path.of("..", "files", dir, fileName + ".txt"));
        } catch (IOException e) {
            throw new RuntimeException("Error reading files", e);
        }
    }

}

File: CSharp/CSharp/Utils/IoUtils.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using CSharp.Model;

namespace CSharp.utils
{
    public static class IoUtils
    {
        public static void OutputToCsv(string fileName, IEnumerable<string> headers, IEnumerable<IEnumerable<string>> rows)
        {

            List<string> fileContents = new List<string>();
            fileContents.Add(String.Join(",", headers));
            fileContents.AddRange(rows.Select(row => String.Join(",", row)));

            var path = $"output/{fileName}.csv";
            
            var file = new System.IO.FileInfo(path);
            if (!file.Directory.Exists)
            {
                file.Directory.Create();
            }

            File.WriteAllLines(
                path,
                fileContents
            );
        }
        
        public static List<string> ReadExpected(int count)
        {
            return GetPath("answers", count).ToList();
        }
        
        public static List<PaperSlip> ReadInput(int count)
        {
            return GetPath("input", count).ToList().Select(PaperSlip.Parse).ToList();
        }

        private static List<string> GetPath(string directory, int count)
        {
            var path = Path.Combine(GetFilesDirPath(), $"./{directory}/{count}.txt");
            return File.ReadAllLines(path).ToList();
        }

        private static string GetFilesDirPath()
        {
            var dirInfo = Directory.GetParent(".");

            while (dirInfo.Name != "three_million_men")
            {
                dirInfo = dirInfo.Parent;
            }

            return dirInfo.GetDirectories().First(dir => dir.Name.Contains("files")).FullName;
        }
    }
}